nixos/modules/services/networking/pihole-ftl.nix

```
@@ -0,0 +1,344 @@
+        # If not specified, defaults to 1 hour.
+        # pihole expects lease time to be in hours.
+        # If infinite is used, set it to 0.
+        DHCP_LEASE_TIME =
+          if (length range) < 4 then 1
+          else
+            let
+              match = builtins.match "([[:digit:]]+)(m|h|d|w)?" (elemAt range 4);
+              multipliers = rec {
+                w = d * 7;
+                d = h * 24;
+                h = 1;
+                m = h / 60.0;
+                s = m / 60.0;
+              };
+              suffix = let value = (elemAt match 1); in if value != null then value else "s";
+              multiplier = getAttr suffix multipliers;
+            in
+            if match != null then (strings.toInt (elemAt match 0)) * multiplier
+            else 0;
+        DHCP_ROUTER = config.networking.defaultGateway.address;
+        DHCP_START = elemAt (splitIPv4 (elemAt range 0)) 6;
+        DHCP_rapid_commit = boolSetting "dhcp-rapid-commit";
+      }
+    )
+    // cfg.extraSetupVars;
+
+  setupVarsConf = (pkgs.formats.keyValue { }).generate "setupVars.conf" setupVars;
+
+  dnsmasqConf = config.services.dnsmasq.configFile;
+
+  # Check the syntax of the config file. Note that this does not check that all options are all valid
+  validatedDnsmasqConf = pkgs.runCommand "validated-dnsmasq.conf" { nativeBuildInputs = [ cfg.package ]; } ''
+    pihole-FTL -- --test --conf-file=${dnsmasqConf} > out 2>&1 || true
+    if ! grep -q "syntax check OK." out; then
+      echo pihole-FTL config validation failed.
+      echo config was ${dnsmasqConf}.
+      echo pihole-FTL output:
+      cat out
+      exit 1
+    fi
+    cp ${dnsmasqConf} $out
+  '';
+
+  pihole-script = pkgs.writeScriptBin "pihole" ''
+    cd ${cfg.package}
+    sudo=exec
+    if [[ "$USER" != pihole ]]; then
+      sudo='exec /run/wrappers/bin/sudo -u pihole'
+    fi
+    $sudo ${pkgs.pihole}/bin/pihole "$@"
+  '';
+in
+{
+  options.services.pihole-ftl = {
+    enable = mkEnableOption (mdDoc "Pi-hole FTL");
+
+    package = mkPackageOptionMD pkgs "pihole-ftl" {};
+
+    privacyLevel = mkOption {
+      type = types.numbers.between 0 3;
+      description = mdDoc ''
+        Level of detail in generated statistics. 0 enables full statistics, 3
+        shows only anonymous statistics.
+
+        See [the documentation](https://docs.pi-hole.net/ftldns/privacylevels).
+
+        Also see services.dnsmasq.settings.log-queries to completely disable
+        query logging.
+      '';
+      default = 0;
+      example = "3";
+    };
+
+    openFirewall = mkOption {
+      type = types.bool;
+      default = true;
+      description = mdDoc "Open ports in the firewall for Pi-hole FTL.";
+    };
+
+
+    stateDirectory = mkOption {
+      type = types.path;
+      default = "/etc/pihole";
+      internal = true;
+      description = ''Path for any pihole state files.
+      pihole does not support any path other than /etc/pihole.'';
```

```suggestion
      description = ''
        Path for any pihole state files.
        pihole does not support any path other than /etc/pihole.
      '';
```
Why is then even an setting? We could just hardcode it